timely/src/dataflow/channels/pact.rs

```diff
@@ -76,6 +76,38 @@ impl<D, F> Debug for Exchange<D, F> {
     }
 }
 
+/// An exchange between multiple observers by data, backed by [LeanExchangePusher].
+pub struct LazyExchange<D, F> { hash_func: F, phantom: PhantomData<D> }
+
+impl<D, F: FnMut(&D)->u64+'static> LazyExchange<D, F> {
+    /// Allocates a new `LeanExchange` pact from a distribution function.
+    pub fn new(func: F) -> Self {
+        Self {
+            hash_func:  func,
+            phantom:    PhantomData,
+        }
+    }
+}
+
+// Exchange uses a `Box<Pushable>` because it cannot know what type of pushable will return from the allocator.
+impl<T: Eq+Data+Clone, D: Data+Clone, F: FnMut(&D)->u64+'static> ParallelizationContract<T, D> for LazyExchange<D, F> {
+    // TODO: The closure in the type prevents us from naming it.
+    //       Could specialize `ExchangePusher` to a time-free version.
+    type Pusher = Box<dyn Push<Bundle<T, D>>>;
+    type Puller = Box<dyn Pull<Bundle<T, D>>>;
+    fn connect<A: AsWorker>(mut self, allocator: &mut A, identifier: usize, address: &[usize], logging: Option<Logger>) -> (Self::Pusher, Self::Puller) {
+        let (senders, receiver) = allocator.allocate::<Message<T, D>>(identifier, address);
+        let senders = senders.into_iter().enumerate().map(|(i,x)| LogPusher::new(x, allocator.index(), i, identifier, logging.clone())).collect::<Vec<_>>();
+        (Box::new(LazyExchangePusher::new(senders, move |_, d| (self.hash_func)(d))), Box::new(LogPuller::new(receiver, allocator.index(), identifier, logging.clone())))
+    }
+}
+
+impl<D, F> Debug for LazyExchange<D, F> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("LeanExchange").finish()
+    }
+}
+
```

Comment:
Should this all be here, or in the `lazy_exchange.rs` file? Ignore the current pacts, which were few enough to not require a new file: probably we should peel off as much as possible and make it all in to a `pub use lazy_exchange;`?

Reply:
Moved the PACT impl to `lazy_exchange.rs`.